### PR TITLE
Introducing delayed reviews

### PR DESCRIPTION
**Pull request by fcamblor**
# New Feature : Delay my review

This PR aims at **introducing a "Delay my review" button** when it comes to evaluate a Proposal (for both Comitee and GT) : 
_CFP__MIR-6146_qweqweqwe_bbb_-_Devoxx_France_2022
_(Note: the reason can be left empty)_

**When a review is marked as "delayed"**, we show the standard "show votes" screen, but without any votes displayed : 
Votes_for_ZEN-3622_test_presentation3_-_Devoxx_France_2022
Votes_for_BST-9767_…_Et_si_on_parlait_santé_mentale_dans_notre_industrie__-_Devoxx_France_2022

# Accessing Delayed Reviews

I provided a dedicated **"Delayed reviews" page** where every pending reviews are displayed with the delayed reason : 
All_my_delayed_reviews_-_Devoxx_France_2022

This page is accessible from various locations : 
- From CFP comitee world : todolist screen, my votes screen, show votes screen
CFP_Admin_-_Devoxx_France_2022
All_my_votes_-_Devoxx_France_2022
Votes_for_ZEN-3622_test_presentation3_-_Devoxx_France_2022

- From GT world : a dedicated navigation menu, GT Home screen, my votes screen, after vote screen
Votes_for_BST-9767_…_Et_si_on_parlait_santé_mentale_dans_notre_industrie__-_Devoxx_France_2022CFP_Admin_-_Devoxx_France_2022All_my_votes_-_Devoxx_France_2022Votes_for_BST-9767_…_Et_si_on_parlait_santé_mentale_dans_notre_industrie__-_Devoxx_France_2022

# Rules about Delayed Reviews

Impacts on existing code are very limited as Delayed Reviews are stored into dedicated collections (`Proposals:DelayedReviews:ByEventCode:DevoxxFr2022:AndAuthor:123456789012345` for CFP comitee, and `ReviewGT:DelayedReviews:ByEventCode:DevoxxFr2022:AndAuthor:123456789012345` for GTs)

The only impacts are : 
- Delayed reviews are excluded from TODO lists (both CFP Comitee & GT ones)
- Delayed reviews are excluded from show votes "next" buttons (we shouldn't navigate to a delayed review from these buttons)
- As soon as a Review is Delayed, we're not showing votes anymore on the show votes screen
- As soon as we vote on a Proposal, DelayedReview (if exists) are removed for targetted proposal (for current user only obviously)

# Side topic : moving watch/unwatch button into the "Evaluate" section

As a side topic, I moved the "watch/unwatch" button into the "Evaluate" section, and put the list of concerned events into a collapsible section : 
_CFP__MIR-6146_qweqweqwe_bbb_-_Devoxx_France_2022
Notification_Center

### CODE REVIEW

File: app/controllers/GoldenTicketController.scala [hunk 1]
```diff
@@ -188,8 +205,12 @@ object GoldenTicketController extends SecureCFPController {
       scala.concurrent.Future {
         Proposal.findById(proposalId) match {
           case Some(proposal) =>
+            val proposalIdsWithDelayedReview = ReviewByGoldenTicket.delayedReviewsReasons(uuid).keySet
```
Comment (nicmarti) [inline app/controllers/GoldenTicketController.scala]: Performance : here you might want to use a lighter Redis call such as Redis HKEYS directly. The HGETALL is a O(N) slow operation (+ the JSON Parsing). 
What about `ReviewByGoldenTicket.allProposalIdsForUser(uuid)` that would return a Set[ProposalID] (or a Set[String])

File: app/controllers/GoldenTicketController.scala [hunk 2]
```diff
@@ -188,8 +205,12 @@ object GoldenTicketController extends SecureCFPController {
...
+            val currentProposalReviewHasBeenDelayed = proposalIdsWithDelayedReview.contains(proposal.id)
```
Comment (nicmarti) [inline app/controllers/GoldenTicketController.scala:209]: You can use Redis.hexists or Redis.exists also

File: app/controllers/GoldenTicketController.scala [hunk 3]
```diff
@@ -237,13 +258,28 @@ object GoldenTicketController extends SecureCFPController {
...
       }.getOrElse {
         BadRequest("Invalid proposal type")
       }
   }
 
+  def delayedReviews() = SecuredAction(IsMemberOfGroups(securityGroups)) {
+    implicit request: SecuredRequest[play.api.mvc.AnyContent] =>
+
+      val uuid = request.webuser.uuid
+      val delayedReviewReasonByProposalId = ReviewByGoldenTicket.delayedReviewsReasons(uuid)
```
Comment (nicmarti) [inline app/controllers/GoldenTicketController.scala:273]: Same as above : here it seems you need only a Set[String] as proposalIds

File: app/models/Review.scala [hunk 4]
```diff
@@ -559,4 +560,24 @@ object Review {
...
+    client =>
+      client.hset(s"Proposals:DelayedReviews:ByEventCode:${ConferenceDescriptor.current().eventCode}:AndAuthor:${uuid}", proposalId, reason)
+  }
+
+  def proposalDelayedReviewReason(uuid: String, proposalId: String): Option[String] = Redis.pool.withClient {
+    client =>
+      client.hget(s"Proposals:DelayedReviews:ByEventCode:${ConferenceDescriptor.current().eventCode}:AndAuthor:${uuid}", proposalId)
+  }
+
+  def delayedReviewsReasons(uuid: String) = Redis.pool.withClient {
```
Comment (nicmarti) [inline app/models/Review.scala]: Please add the return type

File: app/models/ReviewByGoldenTicket.scala [hunk 5]
```diff
@@ -475,4 +476,24 @@ object ReviewByGoldenTicket {
...
+    client =>
+      client.hset(s"ReviewGT:DelayedReviews:ByEventCode:${ConferenceDescriptor.current().eventCode}:AndAuthor:${uuid}", proposalId, reason)
+  }
+
+  def proposalDelayedReviewReason(uuid: String, proposalId: String): Option[String] = Redis.pool.withClient {
+    client =>
+      client.hget(s"ReviewGT:DelayedReviews:ByEventCode:${ConferenceDescriptor.current().eventCode}:AndAuthor:${uuid}", proposalId)
+  }
+
+  def delayedReviewsReasons(uuid: String) = Redis.pool.withClient {
```
Comment (nicmarti) [inline app/models/ReviewByGoldenTicket.scala]: Please add the return type for all public function

File: app/controllers/GoldenTicketController.scala (hunk 1 above)
Comment (fcamblor) [reply app/controllers/GoldenTicketController.scala]: no parsing involved in the `ReviewGT:DelayedReviews:ByEventCode:DevoxxFR2022:AndAuthor:xxxx` collection : this is a simple `proposalId (String) -> reason (String)` map (no JSON serialization/deserialization involved)

But I agree with you on introducing a `allProposalIdsHavingDelayedReviewsForUser(uuid)` in both `Review` & `ReviewByGoldenTicket` models for performance reasons
fixed it through 42055757

File: app/controllers/GoldenTicketController.scala (hunk 2 above)
Comment (fcamblor) [reply app/controllers/GoldenTicketController.scala:209]: In the latest version of `GoldenTicketController.showVotesForProposal()` impl, I need to retrieve total number of delayed reviews (to display it)
=> introducing `allProposalIdsHavingDelayedReviewsForUser` through 42055757 should be enough for this purpose

File: app/models/Review.scala (hunk 4 above)
Comment (fcamblor) [reply app/models/Review.scala]: fixed through f7adcba1

File: app/models/ReviewByGoldenTicket.scala (hunk 5 above)
Comment (fcamblor) [reply app/models/ReviewByGoldenTicket.scala]: fixed through f7adcba1

File: app/controllers/GoldenTicketController.scala (hunk 3 above)
Comment (fcamblor) [reply app/controllers/GoldenTicketController.scala:273]: Disagree on this one as I need to show every delayed reviews' reasons on the screen (this is the main purpose of the delayed review screen : showing all delayed reviews with their reason)

**Review by nicmarti — APPROVED** [review]

File: app/controllers/GoldenTicketController.scala (hunk 1 above)
Comment (nicmarti) [reply app/controllers/GoldenTicketController.scala]: thanks

File: app/controllers/GoldenTicketController.scala (hunk 2 above)
Comment (nicmarti) [reply app/controllers/GoldenTicketController.scala:209]: ok

File: app/controllers/GoldenTicketController.scala (hunk 3 above)
Comment (nicmarti) [reply app/controllers/GoldenTicketController.scala:273]: ok,

**Pull request merged**